src/builtin.js

```diff
@@ -416,111 +282,68 @@ Sk.builtin.max = function max($default, key, args) {
-    if (!Sk.builtin.checkIterable(iter)) {
-        throw new Sk.builtin.TypeError("'" + Sk.abstr.typeName(iter) +
-            "' object is not iterable");
-    }
 
-    it = Sk.abstr.iter(iter);
-    for (i = it.tp$iternext(); i !== undefined; i = it.tp$iternext()) {
+Sk.builtin.any = function any(iter) {
+    const ret = Sk.misceval.iterFor(Sk.abstr.iter(iter), function (i) {
         if (Sk.misceval.isTrue(i)) {
-            return Sk.builtin.bool.true$;
+            return new Sk.misceval.Break(true);
         }
-    }
-
-    return Sk.builtin.bool.false$;
+    });
+    return ret === undefined ? Sk.builtin.bool.false$ : Sk.builtin.bool.true$;
 };
 
-Sk.builtin.all = function all (iter) {
-    var it, i;
-
-    Sk.builtin.pyCheckArgsLen("all", arguments.length, 1, 1);
-    if (!Sk.builtin.checkIterable(iter)) {
-        throw new Sk.builtin.TypeError("'" + Sk.abstr.typeName(iter) +
-            "' object is not iterable");
-    }
-
-    it = Sk.abstr.iter(iter);
-    for (i = it.tp$iternext(); i !== undefined; i = it.tp$iternext()) {
+Sk.builtin.all = function all(iter) {
+    const ret = Sk.misceval.iterFor(Sk.abstr.iter(iter), function (i) {
         if (!Sk.misceval.isTrue(i)) {
-            return Sk.builtin.bool.false$;
+            return new Sk.misceval.Break(false);
         }
-    }
-
-    return Sk.builtin.bool.true$;
+    });
+    return ret === undefined ? Sk.builtin.bool.true$ : Sk.builtin.bool.false$;
 };
 
-Sk.builtin.sum = function sum (iter, start) {
+Sk.builtin.sum = function sum(iter, start) {
     var tot;
-    var intermed;
-    var it, i;
     var has_float;
-
-    Sk.builtin.pyCheckArgsLen("sum", arguments.length, 1, 2);
+    // this follows the order python checks errors
     Sk.builtin.pyCheckType("iter", "iterable", Sk.builtin.checkIterable(iter));
-    if (start !== undefined && Sk.builtin.checkString(start)) {
+    if (Sk.builtin.checkString(start)) {
         throw new Sk.builtin.TypeError("sum() can't sum strings [use ''.join(seq) instead]");
     }
-    if (start === undefined) {
-        tot = new Sk.builtin.int_(0);
-    } else {
-        tot = start;
-    }
-
-    it = Sk.abstr.iter(iter);
-    for (i = it.tp$iternext(); i !== undefined; i = it.tp$iternext()) {
-        if (i instanceof Sk.builtin.float_) {
+    tot = start;
+    Sk.misceval.iterFor(Sk.abstr.iter(iter), function (i) {
+        if (!has_float && i instanceof Sk.builtin.float_) {
             has_float = true;
-            if (!(tot instanceof Sk.builtin.float_)) {
-                tot = new Sk.builtin.float_(Sk.builtin.asnum$(tot));
-            }
-        } else if (i instanceof Sk.builtin.lng) {
-            if (!has_float) {
-                if (!(tot instanceof Sk.builtin.lng)) {
-                    tot = new Sk.builtin.lng(tot);
-                }
-            }
+            tot = new Sk.builtin.float_(Sk.builtin.asnum$(tot));
         }
-
+        // else if (i instanceof Sk.builtin.lng) {
```

Comment:
Yes in py2 mode. This algorithm doesn’t need to check it though since it just uses the nb$add architecture. 

Also long in py2 mode is just a trivial subclass of int_ 
You can now find long  defines at the end of of int.js

Reply:
actually technically this function is already broken.
I'll submit a different pr for this since it's just dealing with floats as being the problem. 
It fails at complex numbers already. 
On master branch:
```python
>>> sum([complex(1,2), 1])
# TypeError: unsupported operand type(s) for +: 'int' and 'complex'
```

Reply:
resolved
see commit 7bbd4b4ef2826831bdc4526174ffe129b80b54c0
and pr #1129